walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,119 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> conflictingTimes = findConflictingTimes(events, request.getAttendees());
+    List<TimeRange> possibleTimes = findPossibleTimes(conflictingTimes, request.getDuration());
+
+    return possibleTimes;
+  }
+
+  public List<TimeRange> findConflictingTimes(Collection<Event> events,
+      Collection<String> attendees) {
+    List<TimeRange> conflicts = findConflicts(events, attendees);
+    List<TimeRange> conflictingTimes = mergeOverlappingRanges(conflicts);
+
+    return conflictingTimes;
+  }
+
+  public List<TimeRange> findConflicts(Collection<Event> events,
+      Collection<String> requestAttendees) {
+    List<TimeRange> conflicts = new ArrayList<>();
+
+    for (Event event : events) {
+      if (!Collections.disjoint(requestAttendees, event.getAttendees())) {
+        conflicts.add(event.getWhen());
+      }
+    }
+
+    Collections.sort(conflicts, TimeRange.ORDER_BY_START);
+
+    return conflicts;
+  }
+
+  public List<TimeRange> mergeOverlappingRanges(List<TimeRange> conflicts) {
+    List<TimeRange> conflictingTimes = new ArrayList<>();
+    TimeRange first = null;
+    TimeRange second = null;
+    TimeRange combined = null;
+    boolean nested = false;
+    boolean useCombinedRange = false;
+
+    for (int i = 0; i < conflicts.size() - 1; i++) {
+      if (useCombinedRange) {
+        first = combined;
+        useCombinedRange = false;
+      } else {
+        first = conflicts.get(i);
+      }
+
+      second = conflicts.get(i + 1);
+      nested = false;
+
+      if (!first.overlaps(second)) {
+        conflictingTimes.add(first);
+      } else if (!first.contains(second)) {
+        combined = TimeRange.fromStartEnd(first.start(), second.end(), false);
+        useCombinedRange = true;
+      } else {
+        conflictingTimes.add(first);
+        nested = true;
+
+        // Skip over adding the second TimeRange
+        i++;
+      }
+    }
+
+    if (conflicts.size() == 0 || nested) {
```

Comment:
nit: the check for the following could have been done earlier in the method (fail fast: It decouples the computational logic from the error logic. By removing the cognitive gap between error checking and handling, it frees up mental processing power. ):

`conflicts.size() == 0`

Reply:
Similarly lines 90-91 could also been checked earlier.